File: sdk/python/src/rocksky/resources/feed.py
"""``app.rocksky.feed.*`` — feeds, recommendations, search, stories."""

from __future__ import annotations

from ..models import (
    Feed,
    FeedGenerator,
    Recommendations,
    RecommendedAlbum,
    RecommendedArtist,
    SearchResults,
    Story,
)
from ._base import Resource, parse_list, parse_model


class FeedResource(Resource):
    async def get(
        self,
        feed: str,
        *,
        limit: int | None = None,
        cursor: str | None = None,
    ) -> Feed:
        """Fetch posts from a feed generator URI."""
        data = await self._transport.query(
            "app.rocksky.feed.getFeed",
            params={"feed": feed, "limit": limit, "cursor": cursor},
        )
        return parse_model(Feed, data)

    async def get_generator(self, feed: str) -> FeedGenerator:
        data = await self._transport.query(
            "app.rocksky.feed.getFeedGenerator", params={"feed": feed}
        )
        return parse_model(FeedGenerator, data)

    async def list_generators(self, *, size: int | None = None) -> list[FeedGenerator]:
        data = await self._transport.query(
            "app.rocksky.feed.getFeedGenerators", params={"size": size}
        )
        return parse_list(FeedGenerator, data, key="feeds")

    async def search(self, query: str) -> SearchResults:
        data = await self._transport.query(
            "app.rocksky.feed.search", params={"query": query}
        )
        return parse_model(SearchResults, data)

    async def stories(self, *, size: int | None = None) -> list[Story]:
        data = await self._transport.query(
            "app.rocksky.feed.getStories", params={"size": size}
        )
        return parse_list(Story, data, key="stories")

    async def recommendations(
        self,
        did: str,
        *,
        limit: int | None = None,
    ) -> Recommendations:
        data = await self._transport.query(
            "app.rocksky.feed.getRecommendations",
            params={"did": did, "limit": limit},
        )
        return parse_model(Recommendations, data)

    async def artist_recommendations(
        self,
        did: str,
        *,
        limit: int | None = None,
    ) -> list[RecommendedArtist]:
        data = await self._transport.query(
            "app.rocksky.feed.getArtistRecommendations",
            params={"did": did, "limit": limit},
        )
        return parse_list(RecommendedArtist, data, key="artists")

    async def album_recommendations(
        self,
        did: str,
        *,
        limit: int | None = None,
    ) -> list[RecommendedAlbum]:
        data = await self._transport.query(
            "app.rocksky.feed.getAlbumRecommendations",
            params={"did": did, "limit": limit},
        )
        return parse_list(RecommendedAlbum, data, key="albums")

File: sdk/ruby/lib/rocksky/resources/feed.rb
module Rocksky
  module Resources
    # `app.rocksky.feed.*` endpoints.
    class Feed < Base
      # Free-text search across the catalogue.
      # `q` accepts a positional or `query:` keyword; both map to the lexicon's
      # `query` parameter.
      def search(q = nil, query: nil)
        term = q || query
        raise ArgumentError, "search needs a query string" if term.nil? || term.empty?

        @http.query("app.rocksky.feed.search", query: term)
      end

      # List feed generators.
      def get_feed_generators(size: nil)
        query("app.rocksky.feed.getFeedGenerators", size: size)
      end

      # Fetch a feed generator by URI.
      def get_feed_generator(feed:)
        query("app.rocksky.feed.getFeedGenerator", feed: feed)
      end

      # Fetch feed contents.
      def get_feed(feed:, limit: nil, cursor: nil)
        query("app.rocksky.feed.getFeed", feed: feed, limit: limit, cursor: cursor)
      end

      # Stories (recent highlights).
      def get_stories(size: nil)
        query("app.rocksky.feed.getStories", size: size)
      end

      # Track recommendations for an actor.
      def get_recommendations(did:, limit: nil)
        query("app.rocksky.feed.getRecommendations", did: did, limit: limit)
      end

      # Artist recommendations for an actor.
      def get_artist_recommendations(did:, limit: nil)
        query("app.rocksky.feed.getArtistRecommendations", did: did, limit: limit)
      end

      # Album recommendations for an actor.
      def get_album_recommendations(did:, limit: nil)
        query("app.rocksky.feed.getAlbumRecommendations", did: did, limit: limit)
      end
    end
  end
end
